### PR TITLE
Add Moderator Decorators & Chaining & OpenAIClient & ConversationHistory State

### PR DESCRIPTION
**Pull request by aacebo**
- add moderator decorators that can be used to implement customized moderation logic (spoke about creating a prototype that uses decorators instead of a moderator class with the team).
- add chaining to AI and Application classes
- add beginning of OpenAIClient http client (spoke about not using the official openai SDK with the team, we are not using it because it is inflexible and doesn't come with typings).
- add conversation history to conversation state (spoke with the team about creating a prototype for this instead of having a static class)

Resolves https://github.com/microsoft/teams-ai/issues/432

### CODE REVIEW

File: python/packages/ai/.pylintrc
Comment (blackchoey) [inline python/packages/ai/.pylintrc:192]: Seems `n` is not used as name in this PR?

File: python/packages/ai/teams/ai/clients/openai/schemas/moderation.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/clients/openai/schemas/moderation.py:19]: May name the function as `from_dict`? This is how SK names the function, and `from_json` is used to initialize from json string.

File: python/packages/ai/teams/ai/clients/openai/schemas/moderation.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/clients/openai/schemas/moderation.py:26]: The dataclass decorator should adds a function to return the dict. Could use that instead of defining our own function.

File: python/packages/ai/teams/ai/clients/openai/schemas/moderation.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/clients/openai/schemas/moderation.py:43]: Same as previous comment, use the function added by dataclass decorator instead of writing our own

File: python/packages/ai/teams/ai/clients/openai/openai_client.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/clients/openai/openai_client.py:42]: If `model` is `None`, will OpenAI return error? Should we remove the `model` property in this case?

File: python/packages/ai/teams/ai/planner/azure_openai_planner_options.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/azure_openai_planner_options.py:19]: Why we need to give it a default value? The folder structure is totally defined by users so we may not make any assumptions to it.

File: python/packages/ai/teams/ai/planner/command_type.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/command_type.py:8]: Can you share why we need to inherit `str`?

File: python/packages/ai/teams/ai/planner/openai_planner.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/openai_planner.py:89]: I have fixed this by using `find()` function in my PR, which is shorter than try/catch block. You can merge my changes after my PR is merged.

File: python/packages/ai/teams/ai/planner/openai_planner.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/openai_planner.py:162]: The `AzureOpenAIPlanner` may not work (I'm not sure), because it inherits `OpenAIPlanner`, and shared the logic here. And even it works, we may not expect SK ensures OpenAITextCompletion always work in the future.

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/openai_planner_options.py:14]: Also why we need to have default prompt_folder?

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/openai_planner_options.py:19]: The constructor is added on purpose to declare some parameter is keyword only, please do not remove it

File: python/packages/ai/teams/ai/planner/planner_error.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/planner/planner_error.py:9]: one guide line we want to follow is do not define new errors if developer cannot handle this error programatically. Can you share how do you expect users catch and handle this error?

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/prompts/prompt_manager.py:35]: Can you share why we need to have default value for prompts_folder?

File: python/packages/ai/teams/ai/ai_options.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/ai_options.py:27]: Per steve's design, prompt manage is now part of planner and not exposed to user. Can remove this option.

File: python/packages/ai/teams/ai/ai_history_options.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/ai_history_options.py:9]: Can you share why we prefer this over an Enum?

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/prompts/prompt_manager.py:26]: Can you share why we need to use generic here? I may need an example to better understand.

File: python/packages/ai/teams/ai/ai.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/ai.py:165]: Same with another comment, since prompt manager is part of planner, maybe we shouldn't add the entrypoint to prompt manager here? Just `planner.add_function()` is fine.

File: python/packages/ai/teams/ai/ai.py
Comment (blackchoey) [inline python/packages/ai/teams/ai/ai.py:255]: Maybe we can use `isinstance` function to detect the command type and finish the type inferring? The type of `plan.commands` can be changed to `Union[PredictedDoCommand, PredictedSayCommand]`. I personally not a fan of current method.

**Comment by blackchoey** [comment]
@aacebo You mentioned a lot of new ideas in PR description, can you create new issues for them with some background (why we want this, etc.)

File: python/packages/ai/.pylintrc
Comment (aacebo) [reply python/packages/ai/.pylintrc:445]: This rule was for "too many branches" which doesn't seem super valuable to me right now since we are porting code that already exists. Later on if we want to remove these and refactor some things we can.

File: python/packages/ai/teams/ai/clients/openai/openai_client.py
Comment (aacebo) [reply python/packages/ai/teams/ai/clients/openai/openai_client.py:42]: if it is done they default to gpt 3.5 I bellieve, it is an optional param in the other SDK's as well. https://github.com/microsoft/teams-ai/blob/main/dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/AI/OpenAI/OpenAIClient.cs#L39

File: python/packages/ai/teams/ai/planner/command_type.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/command_type.py:8]: enums in python default to being `unknown`, if you want the enum to be of type string you need to also inherit from str

File: python/packages/ai/teams/ai/planner/azure_openai_planner_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/azure_openai_planner_options.py:19]: This just provides a suggested default that we use in our samples, I think its useful to cut down on boilerplate as much as possible and devs can always pass in an override in they want to use there own structure. When possible I like to provide default input values that can act as useful examples or for devs that might just want a simple default implementation.

File: python/packages/ai/teams/ai/planner/openai_planner.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/openai_planner.py:162]: When running this code through real testing using the `ListBot`, this was throwing saying it could not find the service `TextCompletionClientBase` and that we were calling the method with too many params. I think something in `SK` isn't working for this method so I changed it to just use the class you are trying to get directly so that we can keep moving forward.

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/openai_planner_options.py:14]: see above comment.

File: python/packages/ai/teams/ai/planner/planner_error.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/planner_error.py:9]: Since our errors inherit they can handle by:

```python
@app.error
async def on_error(context: TurnContext, err: ApplicationError):
    if isinstance(err, PlannerError):
        .. handle planner error
        return
    .. handle normal error
```

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (aacebo) [reply python/packages/ai/teams/ai/prompts/prompt_manager.py:35]: see comment above about defaults

File: python/packages/ai/teams/ai/prompts/prompt_manager_error.py
Comment (aacebo) [reply python/packages/ai/teams/ai/prompts/prompt_manager_error.py:9]: see comment above

File: python/packages/ai/teams/ai/state/message.py
Comment (aacebo) [reply python/packages/ai/teams/ai/state/message.py:9]: this was a suggestion made by Steve, instead of using the plain text conversation history implementation used in the other SDKs (which we will likely change soon) we are using a message object that we based off the one OpenAI uses https://platform.openai.com/docs/guides/gpt so we can more easily query history. DO and SAY are messages that may occur in the conversation history and we want the developers to have an interface for accessing them.

File: python/packages/ai/teams/ai/ai_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai_options.py:27]: because the planner isn't using generics like the rest of the codebase it is throwing type errors so I can't use it for this until it is refactored to use a generic turn state 
image

File: python/packages/ai/teams/ai/ai_history_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai_history_options.py:9]: String literals give you intellisense in the editor as you are working and do not require you to import a class/enum:
image

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (aacebo) [reply python/packages/ai/teams/ai/prompts/prompt_manager.py:26]: Like in the comment about the planner above, if the prompt manager doesn't use the same turn state as the ai/app class it will give type errors. If we want to remove generics we have to do it for the entire project so all the components are compatible. I will look into it today when I get some time.

File: python/packages/ai/teams/ai/ai.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai.py:165]: I think since we have chosen to use decorators anything the user overrides should be accessible via a decorator. This provides a simple decorator for users to register functions the same way they would register an action, so the syntax is consistent.

File: python/packages/ai/teams/ai/ai.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai.py:212]: see above comment about typing errors

File: python/packages/ai/teams/ai/ai.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai.py:255]: good idea I didn't think of that 👍

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/openai_planner_options.py:19]: dataclasses should not have constructors because they provide there own default constructor for the class. The way to initialize fields of a dataclass is via the method above using property initialization.

File: python/packages/ai/teams/ai/actions/action_entry.py
Comment (corinagum) [inline python/packages/ai/teams/ai/actions/action_entry.py:31]: Is it possible to be more specific than `Any` here?

Comment (aacebo) [reply python/packages/ai/teams/ai/actions/action_entry.py:31]: the entities arg can be anything truly since its user defined, we could switch to `Dict[str, Any]` but thats essentially the same thing.

File: python/samples/echo/src/api.py
Comment (corinagum) [inline python/samples/echo/src/api.py:9]: Please follow the samples naming conventions from `js` and `dotnet` (list bot too)

File: python/packages/ai/teams/app.py
Comment (corinagum) [inline python/packages/ai/teams/app.py:8]: Does the python SDK not have `CloudAdapter`? It's the modern version of `BotFrameworkAdapter`

File: python/packages/ai/teams/ai/ai.py
Comment (corinagum) [inline python/packages/ai/teams/ai/ai.py:126]: Line character length seems kind of short to me. I'm inclined to recommend making it longer. `js` is 120, I believe.

File: python/packages/ai/teams/ai/actions/action_entry.py
Comment (corinagum) [reply python/packages/ai/teams/ai/actions/action_entry.py:31]: I'd prefer the later since that's how we do it in `js` but I'll defer to your experience with python

File: python/packages/ai/pyproject.toml
Comment (corinagum) [inline python/packages/ai/pyproject.toml:16]: (for my learning) why does this need to be moved?

**Review by corinagum — APPROVED** [review]
Some nits and questions, LGTM

File: python/packages/ai/pyproject.toml
Comment (aacebo) [reply python/packages/ai/pyproject.toml:16]: I was playing around with using the `openai` sdk so I removed it at some point but ended up readding it later on so it just got moved in that process.

File: python/packages/ai/teams/app.py
Comment (aacebo) [reply python/packages/ai/teams/app.py:8]: ill look and see, all the python examples used this class so I didn't check if the cloud adapter was available. If it exists ill use that instead.

File: python/packages/ai/teams/ai/actions/action_entry.py
Comment (aacebo) [reply python/packages/ai/teams/ai/actions/action_entry.py:31]: I tried but for python that results in type errors since we what to pass class instances in some cases 
image

File: python/packages/ai/teams/app.py
Comment (blackchoey) [inline python/packages/ai/teams/app.py:176]: Seems this should be another feature that JS/C# should add. Can you help create new tasks for JS/C# SDK?

File: python/packages/ai/tests/prompts/test_prompt_manager.py
Comment (blackchoey) [inline python/packages/ai/tests/prompts/test_prompt_manager.py:23]: Thanks for help adding this.
A minor question: I'm just curious about the practice in your mind: should we always add type annotation in our code? You could see some variable declarations does not add type annotation in our code, not sure whether we should update them all.

File: python/packages/ai/teams/ai/planner/azure_openai_planner_options.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/planner/azure_openai_planner_options.py:19]: Thank you, does it make sense to make this parameter as keyword only parameter? If yes, can you move it after `*`?

File: python/packages/ai/teams/ai/planner/command_type.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/planner/command_type.py:8]: Thank you for the explanation. I'm curious why we want to make it to be type string? When we write `CommandType.Do`, do we need to care what's the type of the underlying value?

File: python/packages/ai/teams/ai/planner/openai_planner.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/planner/openai_planner.py:158]: See my comment at https://github.com/microsoft/teams-ai/pull/441#discussion_r1299504736. I would prefer to keep the implementation in my PR because that's how Semantic Kernel handles different (OpenAI & AzureOpenAI) client. Only use OpenAIClient here may cause potential issues when SK releases new package.

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/planner/openai_planner_options.py:14]: Please also consider making it to be a keyword only parameter.

File: python/packages/ai/teams/ai/planner/planner_error.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/planner/planner_error.py:9]: It also comes from Azure SDK guideline (which is summarize to the SDK guideline word I shared with you): do not create a new exception type unless the developer can handle the error programmatically.

This is a minor comment. If you think the guideline makes sense, let's adopt it. If not, we can go ahead with current implementation and discuss the guidelines further.

File: python/packages/ai/teams/ai/state/message.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/state/message.py:9]: Agree that using the data models from OpenAI is better. But the document mentions the available roles are `system`, `user` and `assistant`. So seems `say` and `do` should not in the list?

File: python/packages/ai/teams/ai/ai_options.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/ai_options.py:27]: I see, you're declaring a single StateT instead of using 3 (ConversationT, UserT, TempT), seems mypy cannot handle this.
I'm good to go with current implementation. But I will create new PR to update planner as well as change the implementation to use planner. Let me know if you have any concerns to this.

File: python/packages/ai/teams/ai/ai_history_options.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/ai_history_options.py:9]: Do we need to align all Enums to Literal? How do we decide when to use Literal and when to use Enum?

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/prompts/prompt_manager.py:26]: I'm good to go with your current implementation. But I wrote some examples here: https://github.com/microsoft/teams-ai/blob/1c886a283043b6bd2c1cee1455f08dfce3662c1a/python/packages/ai/examples/openai_planner.py#L45-L61. Not sure how you feels about it. It indeed introduces some difference with other SDK.

File: python/packages/ai/teams/ai/ai.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/ai.py:165]: I‘m neutral to this. Can you create tasks for JS/C# to add this? It should be doable in all SDKs and we can drive this if thinking it's valuable to users.

**Comment by blackchoey** [comment]
Approving first. Please resolve the comments remained open before merge.

File: python/packages/ai/tests/prompts/test_prompt_manager.py
Comment (aacebo) [reply python/packages/ai/tests/prompts/test_prompt_manager.py:23]: that is a really good question that I have also thought about, We should decide together what practice is best for our team so that we are both on the same page. I _personally_ do the following: use explicit type annotations whenever exposing a type to the user ie options/class properties. I usually use inferred in cases like the one above where its internal only code, or if the language server for some reason gets it wrong which rarely but sometimes happens.

Lets discuss this in more detail during our weekly sync and if we decide to move towards using explict types everywhere I can open a PR to update the codebase and add linting to support it.

File: python/packages/ai/teams/ai/prompts/prompt_manager.py
Comment (aacebo) [reply python/packages/ai/teams/ai/prompts/prompt_manager.py:26]: That makes sense to me, I'm going to create a task to handle that work separately but I will send it to you and start it today so that all our state uses that method.

File: python/packages/ai/teams/ai/ai_history_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai_history_options.py:9]: My _personal opinion_ is that literals should be used for public types so that our users get intellisense and doesn't have to import another class from the library, thus making boilerplate smaller and easier to read. I am also for using literals everywhere in the codebase for this same reason, so if you think it makes sense to change all our enums to literal string types I support that and I can add it as a cleanup work item.

File: python/packages/ai/teams/ai/ai_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/ai_options.py:27]: I think what you said below about using the three generics makes sense and if you are in aligment we can merge this PR as is for now and Ill open a new one to change all our `StateT` generics to use your method (3 generics) instead.

File: python/packages/ai/teams/ai/state/message.py
Comment (aacebo) [reply python/packages/ai/teams/ai/state/message.py:9]: Ya thats our own custom stuff because the `ConversationHistory` class needs to be able to do things like "get the last `say` message from history and replace it". I created this `Message` class to help do these queries in a more reliable way than using an array of strings like the JS implementation does because after talking with Steve he said we intend to make a similar change in the other libraries anyway. 

Do you have any ideas how we could do this better than what I've added in this PR? I am open to alternative approaches and maybe we can discuss this more in our weekly sync. If we come up with something better we can create a work item to improve upon or redo this history class entirely.

File: python/packages/ai/teams/ai/planner/planner_error.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/planner_error.py:9]: Ok that works for me, In this case would we just have one error type `ApplicationError` or should we use a builtin error like `ValueError`? I am going to merge this today but during our sync we can discuss this more and create a work item to fix how we are handling errors in the codebase.

Comment (aacebo) [reply python/packages/ai/teams/ai/planner/planner_error.py:9]: When implementing this I was following the C# implementation but if the python way of doing things is to use builtin errors more then I'm fully behind doing it that way 👍

File: python/packages/ai/teams/ai/planner/openai_planner_options.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/openai_planner_options.py:19]: ok thanks for explaining that to me in more detail. It makes sense to me if we do what you said earlier actually and change these `Options` classes to just be constructor arguments instead, and the optional ones can just be positional args. Ill create a work item to track that work that will take place in a separate PR.

File: python/packages/ai/teams/ai/planner/command_type.py
Comment (aacebo) [reply python/packages/ai/teams/ai/planner/command_type.py:8]: Ya its just good to know if an enum is a `string` enum or a `int` enum at least from an intellisense perspective. Not necessarily _needed_ since typing in python is optional anyway, but I personally like the types to be as accurate as possible especially in cases like these that don't require alot of work.

File: python/packages/ai/teams/ai/state/message.py
Comment (blackchoey) [reply python/packages/ai/teams/ai/state/message.py:9]: Thank you for the explanation. I fine with it now by knowing the reason of having "say" and "do".

**Pull request merged**
